sky/data/sky_csync.py

```diff
@@ -0,0 +1,337 @@
+    _CONN.commit()
+
+
+@connect_db
+def _get_all_running_csync_pid() -> List[Any]:
+    """Returns all the registerd pid of CSYNC processes"""
+    assert _CURSOR is not None
+    _CURSOR.execute('SELECT csync_pid FROM running_csync WHERE boot_time=(?)',
+                    (_BOOT_TIME,))
+    rows = _CURSOR.fetchall()
+    csync_pids = [row[0] for row in rows]
+    return csync_pids
+
+
+@connect_db
+def _set_running_csync_sync_pid(csync_pid: int, sync_pid: Optional[int]):
+    """Given the process id of CSYNC, sets the sync_pid column value"""
+    assert _CURSOR is not None
+    assert _CONN is not None
+    _CURSOR.execute(
+        'UPDATE running_csync '
+        'SET sync_pid=(?) WHERE csync_pid=(?)', (sync_pid, csync_pid))
+    _CONN.commit()
+
+
+@connect_db
+def _get_running_csync_sync_pid(csync_pid: int) -> Optional[int]:
+    """Given the process id of CSYNC, returns the sync_pid column value"""
+    assert _CURSOR is not None
+    _CURSOR.execute(
+        'SELECT sync_pid FROM running_csync '
+        'WHERE csync_pid=(?) AND boot_time=(?)', (csync_pid, _BOOT_TIME))
+    row = _CURSOR.fetchone()
+    if row:
+        return row[0]
+    raise ValueError(f'CSYNC PID {csync_pid} not found.')
+
+
+@connect_db
+def _delete_running_csync(csync_pid: int):
+    """Deletes the row with process id of CSYNC from running_csync table"""
+    assert _CURSOR is not None
+    assert _CONN is not None
+    _CURSOR.execute('DELETE FROM running_csync WHERE csync_pid=(?)',
+                    (csync_pid,))
+    _CONN.commit()
+
+
+@connect_db
+def _get_csync_pid_from_source_path(path: str) -> Optional[int]:
+    """Given the path, returns process ID of csync running on it"""
+    assert _CURSOR is not None
+    _CURSOR.execute(
+        'SELECT csync_pid FROM running_csync '
+        'WHERE source_path=(?) AND boot_time=(?)', (path, _BOOT_TIME))
+    row = _CURSOR.fetchone()
+    if row:
+        return row[0]
+    return None
+
+
+@click.group()
+def main():
+    pass
+
+
+def get_s3_upload_cmd(src_path: str, dst: str, num_threads: int, delete: bool,
+                      no_follow_symlinks: bool):
+    """Builds sync command for aws s3"""
+    config_cmd = ('aws configure set default.s3.max_concurrent_requests '
+                  f'{num_threads}')
+    subprocess.check_output(config_cmd, shell=True)
+    sync_cmd = f'aws s3 sync {src_path} s3://{dst}'
+    if delete:
+        sync_cmd += ' --delete'
+    if no_follow_symlinks:
+        sync_cmd += ' --no-follow-symlinks'
+    return sync_cmd
+
+
+def get_gcs_upload_cmd(src_path: str, dst: str, num_threads: int, delete: bool,
+                       no_follow_symlinks: bool):
+    """Builds sync command for gcp gcs"""
+    sync_cmd = (f'gsutil -m -o \'GSUtil:parallel_thread_count={num_threads}\' '
+                'rsync -r')
+    if delete:
+        sync_cmd += ' -d'
+    if no_follow_symlinks:
+        sync_cmd += ' -e'
+    sync_cmd += f' {src_path} gs://{dst}'
+    return sync_cmd
```

Comment:
Also, for discussion, maybe we could use some derivative to make the code cleaner? like
```
class CSYNCCommand:
    # TODO: Add type annotation to arguments
    @classmethod
    def upload_cmd(cls, src, dst, num_threads, delete, no_follow_symlinks) -> str:
        cmd = cls._PREFIX
        if delete:
            cmd.append(cls._DELETE_OPTION)
        if no_follow_symlinks:
            cmd.append(cls._NO_FOLLOW_SYMLINKS_OPTION)
        cmd += [src, cls.format_dst(dst)]
        return ' '.join(cmd)
    @classmethod
    def config_cmd(cls) -> Optional[str]:
        raise NotImplementedError
    @classmethod
    def format_dst(cls, dst: str) ->str:
        raise NotImplementedError
class S3CSYNCCommand(CSYNCCommand):
    _PREFIX = ['aws', 's3', 'sync']
    _DELETE_OPTION = '--delete'
    _NO_FOLLOW_SYMLINKS_OPTION = '--no-follow...'
    @classmethod
    def config_cmd(cls) -> Optional[str]:
        return 'set max req...'
    @classmethod
    def format_dst(cls, dst: str) ->str:
        return f's3://{dst}'
class GCSCSYNCCommand(CSYNCCommand):
    @classmethod
    def config_cmd(cls) -> Optional[str]:
        return None
    @classmethod
    def format_dst(cls, dst: str) ->str:
        return f'gs://{dst}'
```
Not 100% sure. we should discuss which is better.

Reply:
For this PR, `CSYNC` for `IBM` and `R2` is not supported. Added a check in `storage.py/add_store`.

Reply:
Regards to implementing a class, I agree that it looks cleaner with the abstraction. But I'm a bit concerned as well on how shallow the class is since there isn't much going on besides building simple commands. And shallow classes can add more complexity to the project. I'm wondering if it might be better to postpone this implementation until we have more going on with the implementation? For cleaner code, I refactored `get_s3_upload_cmd`/`get_gcs_upload_cmd` from concatenating strings to list append.